projects/node-http-exercise/node-http-exercise.md

```diff
@@ -0,0 +1,255 @@
+
+In order to fully test your API, you will be using Postman to test your routes. For now, what you will be doing is to create the shell of your API by establishing the logic used in your server. There are many ways to create a server. In this assignment, you will be guided in creating a certain type of server logic, but you could as easily use simple `if else` statements or even a `switch` method to render your data. However, since there are quite a few routes required in this API, let's have a structure that will simplify the code. You will have a cleaner API and it will be easily scalable for your own projects. The logic will be simple: we will construct an object containing the routes that will shape the response according to the API requirements, and a default response for routes that are not defined.
+
+1. The first thing you should do is to observe the `req.url` object. For this, we will be using the `parse()` method in the _url_ module you required at the top. I invite you to type in this code in your server function:
+
+```javascript
+let path = url.parse(req.url, true);
+console.log(path);
+res.writeHead(200, 'OK', { 'Content-Type': 'text/plain' });
+res.end();
+```
+
+and then run the server and simply open a browser window and go to `http://localhost:5000`. When you come back to your terminal, you should see something that looks like this:
+
+```bash
+Listening on port 5000...
+Url {
+  protocol: null,
+  slashes: null,
+  auth: null,
+  host: null,
+  port: null,
+  hostname: null,
+  hash: null,
+  search: null,
+  query: [Object: null prototype] {},
+  pathname: '/',
+  path: '/',
+  href: '/'
+}
+```
+
+These are the parameters you will be using to build your API. Worthy of note are `pathname`, `path`, and `query`.
+
+Back to creating the logic of your server, now. When the client makes a request to an endpoint, you want to examine what that endpoint is and then act according to what your API should do. So, for that, we will be creating a helper function that will parse the url to fetch only the pathname, then look into an object containing all your routes to see if this route exists, then build your response accordingly.
+
+There are many ways to do this. Some developers use Regex, others use parsing libraries (such as `utils`), but for the purpose of this assignment, we will be using plain old JavaScript.
+
+2. Within the server, create default status code as `404` (not found).
+
+3. Then, still using any variable name you want, create default content-type as `application/json`.
+
+4. Set a default response object as such:
+
+```javascript
+let response = {
+  message: 'Not Found',
+};
+```
+
+5. Set an empty header object:
+
+```javascript
+let header = {};
+```
+
+6. Refactor the writing header function to use the placeholders:
+
+```javascript
+res.writeHead(status, { 'Content-Type': contentType });
+```
+
+7. Use the `JSON.stringify()` function to parse the response object and send it back in your `.end()`:
```

Comment:
I would suggest they try it first without stringifying it and then they can learn to read the error message.

Reply:
fixed in [799330c](https://github.com/Techtonica/curriculum/pull/1406/commits/799330cf2b82a0e0d74d3cd2e258f97adc3e70db)